File: cmd/goshcoder/sessionpicker.go
package main

// The resume picker: choosing a session at startup with -resume, and switching
// to one mid-session with /resume.

import (
	"bufio"
	"context"
	"errors"
	"fmt"
	"os"
	"os/exec"
	"path/filepath"
	"strconv"
	"strings"
	"sync"
	"time"

	"goshcoder/internal/sessionlog"
)

// pickerLimit caps how many sessions a picker offers. Beyond this the search
// box is the tool, not the list.
const pickerLimit = 50

// listSessionsForPicker gathers the candidates for a picker.
//
// Transcript text is loaded, not just metadata. Nobody remembers a session by
// its first message; they remember "the one where we fixed the SSE reader". A
// metadata-only search is useful for "yesterday's" and useless for "that one
// from October". sessionlog bounds the text it keeps per session so this stays
// affordable.
func listSessionsForPicker(workdir string, allWorkspaces bool) ([]sessionlog.Info, error) {
	store := activeStore()
	sessions, err := store.List(workdir, sessionlog.ListOptions{
		AllWorkspaces: allWorkspaces,
		Limit:         pickerLimit,
		WithText:      true,
	})
	if err != nil {
		return nil, err
	}
	if len(sessions) > 0 || allWorkspaces {
		return sessions, nil
	}
	// A session started at the repository root is not in this subdirectory's
	// shard. Looking there is what makes `chat -resume` work from anywhere in a
	// checkout rather than only from where the session began.
	if root := repositoryRoot(workdir); root != "" && root != workdir {
		return store.List(root, sessionlog.ListOptions{Limit: pickerLimit, WithText: true})
	}
	return nil, nil
}

// repositoryRoot returns the enclosing git worktree root, or "".
func repositoryRoot(dir string) string {
	// One bounded subprocess, matching the sidebar's git inspection: an
	// unbounded one on a network filesystem would hang startup.
	ctx, cancel := context.WithTimeout(context.Background(), 2*time.Second)
	defer cancel()
	command := exec.CommandContext(ctx, "git", "-C", dir, "rev-parse", "--show-toplevel")
	output, err := command.Output()
	if err != nil {
		return ""
	}
	root := strings.TrimSpace(string(output))
	if root == "" {
		return ""
	}
	// Session shards are keyed by the resolved path, so a symlinked checkout
	// must be resolved the same way here or the shard is never found.
	if resolved, err := filepath.EvalSymlinks(root); err == nil {
		return resolved
	}
	return root
}

// matchesSession reports whether every search term appears somewhere in the
// session's identity or its transcript.
//
// Multi-term matching, copied from the /model picker: it is the only existing
// matcher that tolerates whitespace, and a search box that breaks on a space is
// worse than no search box.
func matchesSession(info sessionlog.Info, terms []string) bool {
	haystack := strings.ToLower(strings.Join([]string{
		info.ID, info.Name, info.FirstMessage, info.Cwd, info.SearchText,
	}, " "))
	for _, term := range terms {
		if !strings.Contains(haystack, term) {
			return false
		}
	}
	return true
}

// describeSession renders one picker row. label is the id prefix to show,
// which the caller widens with sessionlog.ShortIDs so the rows stay
// distinguishable.
func describeSession(info sessionlog.Info, showCwd bool) (label, description string) {
	return describeSessionAs(info, info.ShortID(), showCwd)
}

func describeSessionAs(info sessionlog.Info, label string, showCwd bool) (string, string) {
	parts := []string{info.Modified.Local().Format("Jan 2 15:04")}
	if info.Messages > 0 {
		parts = append(parts, fmt.Sprintf("%d msg", info.Messages))
	}
	if info.Locked {
		parts = append(parts, "open elsewhere")
	}
	if showCwd {
		parts = append(parts, filepath.Base(info.Cwd))
	}
	title := info.Title()
	if title != "" && title != info.ID {
		parts = append(parts, title)
	}
	return label, strings.Join(parts, " · ")
}

// chooseSessionInteractively runs the startup picker for -resume.
//
// It runs before the fullscreen program starts, so it uses the same
// line-oriented selection the model onboarding uses rather than a Bubble Tea
// view of its own.
func chooseSessionInteractively(workdir string) (sessionlog.Info, bool, error) {
	sessions, err := listSessionsForPicker(workdir, false)
	if err != nil {
		return sessionlog.Info{}, false, err
	}
	if len(sessions) == 0 {
		fmt.Fprintln(os.Stderr, dim("no saved sessions for this workspace; starting a new one"))
		return sessionlog.Info{}, false, nil
	}

	fmt.Fprintln(os.Stderr, dim("saved sessions for this workspace:"))
	labels := sessionlog.ShortIDs(sessions)
	for index, info := range sessions {
		label, description := describeSessionAs(info, labels[index], false)
		fmt.Fprintf(os.Stderr, "%3d  %s  %s\n", index+1, label, dim(description))
	}
	fmt.Fprintf(os.Stderr, "%s\n", dim("choose a number, or press Enter to start a new session:"))

	choice, err := readLineFromStdin()
	if err != nil {
		return sessionlog.Info{}, false, err
	}
	choice = strings.TrimSpace(choice)
	if choice == "" {
		return sessionlog.Info{}, false, nil
	}
	index, err := parsePositiveIndex(choice, len(sessions))
	if err != nil {
		return sessionlog.Info{}, false, err
	}
	return sessions[index-1], true, nil
}

// switchSession swaps the live session for another one mid-conversation.
//
// The target is attached BEFORE the current file is closed. Closing first left
// every failure path -- a claim held elsewhere, a pre-v3 file, an I/O error --
// with no recorder at all: the window kept running and silently recorded
// nothing for the rest of its life.
//
// The recorder object is reused rather than replaced. Agent.Subscribe binds its
// listener to the recorder it was given, so installing a new one would leave
// every event going to the old, closed recorder.
func (s *session) switchSession(ref string) error {
	if s == nil {
		return errors.New("no session")
	}
	if s.log == nil {
		return errors.New("this session is not being saved, so it cannot be switched")
	}
	if s.agent.State().IsStreaming {
		return errors.New("wait for the current response to finish before switching sessions")
	}
	store := s.sessionStore()
	workdir := s.workspaceRoot()
	info, err := store.Resolve(workdir, ref)
	if err != nil {
		return err
	}
	if info.Path == s.log.path() {
		return errors.New("that is already the current session")
	}

	writer, report, err := store.Attach(info.Path)
	if err != nil {
		if errors.Is(err, sessionlog.ErrBusy) {
			return fmt.Errorf("%s is open in another window (%s)", info.ShortID(), info.Owner)
		}
		if errors.Is(err, sessionlog.ErrLegacyFormat) {
			return fmt.Errorf("%s is an older pi session; start it with -session %s to copy it into a continuable one",
				info.ShortID(), info.ShortID())
		}
		return err
	}

	projected := projectSession(writer.Snapshot(), writer.Leaf())
	// Seeding the transcript before the swap keeps the agent and the log
	// describing the same conversation at every instant.
	s.agent.SetMessages(projected.Messages)
	s.contextEstimate.reset()

	if previous := s.log.swap(writer); previous != nil {
		if closeErr := previous.Close(); closeErr != nil {
			s.pushNotice("Session", "the previous session did not close cleanly: "+closeErr.Error())
		}
	}

	for _, notice := range reportNotices(report) {
		s.pushNotice("Session", notice)
	}
	for _, warning := range projected.Warnings {
		s.pushNotice("Session", warning)
	}
	if projected.Provider != "" && projected.ModelID != "" {
		if err := s.setModel(projected.Provider + "/" + projected.ModelID); err != nil {
			s.pushNotice("Session", fmt.Sprintf(
				"this session was held on %s/%s, which is not available now: %v",
				projected.Provider, projected.ModelID, err))
		}
	}
	s.pushNotice("Session", fmt.Sprintf("switched to %s (%d messages)", info.ShortID(), len(projected.Messages)))
	return nil
}

// sessionListLines renders /sessions.
//
// It is deliberately short. runFullscreenCommand folds a command's whole output
// into one capped message card, so a fifty-row listing arrives as a single
// unscrollable blob; the picker is the real interface and the CLI is where a
// long listing belongs.
func sessionListLines(workdir string) []string {
	sessions, err := listSessionsForPicker(workdir, false)
	if err != nil {
		return []string{"could not read sessions: " + err.Error()}
	}
	if len(sessions) == 0 {
		return []string{"no saved sessions for this workspace"}
	}
	const shown = 10
	labels := sessionlog.ShortIDs(sessions)
	lines := make([]string, 0, shown+1)
	for index, info := range sessions {
		if index >= shown {
			lines = append(lines, fmt.Sprintf("… %d more · goshcoder sessions list", len(sessions)-shown))
			break
		}
		label, description := describeSessionAs(info, labels[index], false)
		lines = append(lines, label+"  "+description)
	}
	lines = append(lines, "/resume <search> to switch")
	return lines
}

// fullscreenResumeSuggestions builds the /resume drill-down.
func fullscreenResumeSuggestions(sessions []sessionlog.Info, query string, currentPath string) []fullscreenSuggestion {
	terms := strings.Fields(strings.ToLower(strings.TrimSpace(query)))
	suggestions := make([]fullscreenSuggestion, 0, len(sessions))
	labels := sessionlog.ShortIDs(sessions)
	for index, info := range sessions {
		// Guarded on a non-empty currentPath: an unsaved session has none, and
		// an empty match would hide every candidate.
		if currentPath != "" && info.Path == currentPath {
			continue
		}
		if len(terms) > 0 && !matchesSession(info, terms) {
			continue
		}
		label, description := describeSessionAs(info, labels[index], false)
		suggestions = append(suggestions, fullscreenSuggestion{
			label:       label,
			description: description,
			value:       "/resume " + label,
			execute:     true,
		})
	}
	return suggestions
}

// cachedSessionSuggestions keeps the picker off the render path.
//
// suggestions() runs inside View(), on every keystroke and on a 120 ms tick.
// Caching alone was not enough: the very first "/resume " keystroke still
// populated the cache inline, and populating it parses every session file in
// the shard and can fork git for the repository root. On a large history that
// is seconds of frozen interface, and on a network filesystem it can hang.
//
// The scan therefore runs on its own goroutine. Until it lands the picker
// simply reports that it is still loading, and invalidation clears the flag so
// the next miss starts a fresh scan.
func (model *fullscreenModel) cachedSessionSuggestions(query string) []fullscreenSuggestion {
	if !model.sessionChoicesValid {
		if !model.sessionScanRunning {
			model.sessionScanRunning = true
			root := model.session.workspaceRoot()
			target := model.sessionScanResult
			go func() {
				sessions, err := listSessionsForPicker(root, false)
				if err != nil {
					sessions = nil
				}
				select {
				case target <- sessions:
				default:
				}
			}()
		}
		select {
		case sessions := <-model.sessionScanResult:
			model.sessionChoices = sessions
			model.sessionChoicesValid = true
			model.sessionScanRunning = false
		default:
			return []fullscreenSuggestion{{
				label: "…", description: "reading saved sessions", value: "", execute: false,
			}}
		}
	}
	current := ""
	if model.session != nil && model.session.log != nil {
		current = model.session.log.path()
	}
	return fullscreenResumeSuggestions(model.sessionChoices, query, current)
}

// invalidateSessionSuggestions forces the picker to be rebuilt.
func (model *fullscreenModel) invalidateSessionSuggestions() {
	model.sessionChoicesValid = false
}

// missingWorkspaceNotice warns when a resumed session's workspace is gone.
//
// Without it the first tool call fails inside os.OpenRoot with an error that
// names neither the session nor the directory it expected.
func missingWorkspaceNotice(info sessionlog.Info, workdir string) string {
	if info.Cwd == "" || info.Cwd == workdir {
		return ""
	}
	if _, err := os.Stat(info.Cwd); err == nil {
		return fmt.Sprintf("this session was started in %s; tools will operate on %s instead", info.Cwd, workdir)
	}
	return fmt.Sprintf("this session's original workspace %s no longer exists; tools will operate on %s instead",
		info.Cwd, workdir)
}

// stdinReader is shared by every prompt that reads a line before the
// interactive interface starts.
//
// A fresh bufio.Reader per prompt discards whatever it buffered past the line
// it returned, so a second prompt -- or the chat loop itself -- silently lost
// input the user had already typed.
var (
	stdinOnce   sync.Once
	stdinShared *bufio.Reader
)

func sharedStdin() *bufio.Reader {
	stdinOnce.Do(func() { stdinShared = bufio.NewReader(os.Stdin) })
	return stdinShared
}

// readLineFromStdin reads one line for the startup picker.
func readLineFromStdin() (string, error) {
	line, err := sharedStdin().ReadString('\n')
	if err != nil && line == "" {
		return "", err
	}
	return line, nil
}

// parsePositiveIndex reads a 1-based menu choice.
func parsePositiveIndex(input string, limit int) (int, error) {
	index, err := strconv.Atoi(input)
	if err != nil || index < 1 || index > limit {
		return 0, fmt.Errorf("choose a number between 1 and %d", limit)
	}
	return index, nil
}

File: src/session_picker.rs
//! Shared, line-oriented session selection for interactive chat startup.
//!
//! The picker deliberately runs before either terminal frontend owns the
//! screen. Keeping its input/output injected makes it usable for Ratatui,
//! line-mode chat, and tests without coupling session discovery to a UI loop.

use std::{
    error::Error,
    io::{self, BufRead, Write},
    path::{Path, PathBuf},
    process::{Command, Stdio},
    time::{SystemTime, UNIX_EPOCH},
};

use time::{OffsetDateTime, macros::format_description};

use crate::sessionlog::{self, ListOptions, SessionInfo, Store};

/// Bounds the startup choice list; transcript search exists for older work.
pub const PICKER_LIMIT: usize = 50;

const PICKER_TIME_FORMAT: &[time::format_description::FormatItem<'static>] =
    format_description!("[month repr:short] [day padding:none] [hour padding:zero]:[minute]");

/// Lists the sessions suitable for `chat -resume`.
///
/// Transcript text is included so future interactive filtering can identify a
/// session by the work discussed in it instead of only its metadata. If the
/// current directory has no sessions, retry at the enclosing repository root:
/// sessions are sharded by workspace and chat is commonly launched from a
/// subdirectory.
pub fn list_sessions_for_picker(
    store: &Store,
    workdir: &Path,
) -> Result<Vec<SessionInfo>, Box<dyn Error>> {
    let options = ListOptions {
        limit: PICKER_LIMIT,
        with_text: true,
        ..ListOptions::default()
    };
    let sessions = store.list(workdir, options)?;
    if !sessions.is_empty() {
        return Ok(sessions);
    }

    let canonical_workdir = workdir
        .canonicalize()
        .unwrap_or_else(|_| workdir.to_path_buf());
    let Some(root) = repository_root(workdir) else {
        return Ok(sessions);
    };
    if root == canonical_workdir {
        return Ok(sessions);
    }
    Ok(store.list(root, options)?)
}

/// Returns the enclosing Git worktree root when it can be determined.
///
/// Failure to inspect Git must never prevent a new chat session from opening,
/// so this intentionally treats an unavailable or non-repository directory as
/// no fallback rather than surfacing an error.
pub fn repository_root(workdir: &Path) -> Option<PathBuf> {
    let output = Command::new("git")
        .arg("-C")
        .arg(workdir)
        .args(["rev-parse", "--show-toplevel"])
        .stdin(Stdio::null())
        .stderr(Stdio::null())
        .output()
        .ok()?;
    if !output.status.success() {
        return None;
    }
    let root = String::from_utf8(output.stdout).ok()?;
    let root = root.trim();
    if root.is_empty() {
        return None;
    }
    let root = PathBuf::from(root);
    Some(root.canonicalize().unwrap_or(root))
}

/// Tests whether every normalized search term occurs in a session's metadata
/// or bounded transcript text.
pub fn matches_session(info: &SessionInfo, terms: &[String]) -> bool {
    let haystack = [
        info.id.as_str(),
        info.name.as_str(),
        info.first_message.as_str(),
        info.cwd.as_str(),
        info.search_text.as_str(),
    ]
    .join(" ")
    .to_ascii_lowercase();
    terms.iter().all(|term| haystack.contains(term))
}

/// Returns the readable label and metadata shown for one session.
pub fn describe_session(info: &SessionInfo, label: &str, show_cwd: bool) -> (String, String) {
    let mut parts = vec![format_modified(info.modified)];
    if info.messages > 0 {
        parts.push(format!("{} msg", info.messages));
    }
    if info.locked {
        parts.push("open elsewhere".to_owned());
    }
    if show_cwd {
        let cwd = Path::new(&info.cwd);
        if let Some(name) = cwd.file_name().filter(|name| !name.is_empty()) {
            parts.push(name.to_string_lossy().into_owned());
        }
    }
    let title = info.title();
    if !title.is_empty() && title != info.id {
        parts.push(title.to_owned());
    }
    (label.to_owned(), parts.join(" · "))
}

/// Renders the startup picker and returns the selected session, if any.
///
/// Pressing Enter is an intentional request for a fresh session. The caller
/// owns configuration mutation and converts a selected ID into the normal
/// `SessionSelection::Session` lifecycle path.
pub fn choose_session<R: BufRead, W: Write>(
    store: &Store,
    workdir: &Path,
    input: &mut R,
    output: &mut W,
) -> Result<Option<SessionInfo>, Box<dyn Error>> {
    let sessions = list_sessions_for_picker(store, workdir)?;
    choose_from_sessions(&sessions, input, output)
}

fn choose_from_sessions<R: BufRead, W: Write>(
    sessions: &[SessionInfo],
    input: &mut R,
    output: &mut W,
) -> Result<Option<SessionInfo>, Box<dyn Error>> {
    if sessions.is_empty() {
        writeln!(
            output,
            "no saved sessions for this workspace; starting a new one"
        )?;
        return Ok(None);
    }

    writeln!(output, "saved sessions for this workspace:")?;
    let labels = sessionlog::short_ids(sessions);
    for (index, (session, label)) in sessions.iter().zip(labels).enumerate() {
        let (label, description) = describe_session(session, &label, false);
        writeln!(output, "{:>3}  {label}  {description}", index + 1)?;
    }
    writeln!(
        output,
        "choose a number, or press Enter to start a new session:"
    )?;
    output.flush()?;

    let mut choice = String::new();
    if input.read_line(&mut choice)? == 0 {
        return Err(io::Error::new(
            io::ErrorKind::UnexpectedEof,
            "session picker input ended before a choice was made",
        )
        .into());
    }
    let choice = choice.trim();
    if choice.is_empty() {
        return Ok(None);
    }
    let selected = choice.parse::<usize>().map_err(|_| {
        io::Error::new(
            io::ErrorKind::InvalidInput,
            format!("choose a number between 1 and {}", sessions.len()),
        )
    })?;
    if !(1..=sessions.len()).contains(&selected) {
        return Err(io::Error::new(
            io::ErrorKind::InvalidInput,
            format!("choose a number between 1 and {}", sessions.len()),
        )
        .into());
    }
    Ok(Some(sessions[selected - 1].clone()))
}

fn format_modified(modified: SystemTime) -> String {
    let seconds = modified
        .duration_since(UNIX_EPOCH)
        .map_or(0, |duration| duration.as_secs().min(i64::MAX as u64) as i64);
    OffsetDateTime::from_unix_timestamp(seconds)
        .ok()
        .and_then(|time| time.format(PICKER_TIME_FORMAT).ok())
        .unwrap_or_else(|| "unknown time".to_owned())
}

#[cfg(test)]
mod tests {
    use std::io::Cursor;

    use super::*;
    use crate::sessionlog::LockOwner;

    fn session(id: &str, name: &str, first_message: &str, search_text: &str) -> SessionInfo {
        SessionInfo {
            id: id.to_owned(),
            path: PathBuf::from(format!("/sessions/{id}.jsonl")),
            cwd: "/work/project".to_owned(),
            name: name.to_owned(),
            first_message: first_message.to_owned(),
            created: None,
            modified: UNIX_EPOCH,
            messages: 4,
            cleared: 0,
            size: 0,
            search_text: search_text.to_owned(),
            locked: false,
            owner: LockOwner::default(),
        }
    }

    #[test]
    fn transcript_search_requires_each_term() {
        let matching = session(
            "aaaa1111-0000-7000-8000-000000000000",
            "SSE fix",
            "hello",
            "we fixed the streaming reader and retry loop",
        );
        let unrelated = session(
            "bbbb2222-0000-7000-8000-000000000000",
            "other",
            "unrelated",
            "tax returns",
        );
        let terms = ["streaming", "retry"]
            .into_iter()
            .map(str::to_owned)
            .collect::<Vec<_>>();

        assert!(matches_session(&matching, &terms));
        assert!(!matches_session(&unrelated, &terms));
    }

    #[test]
    fn picker_renders_unique_labels_and_returns_selected_session() {
        let sessions = vec![
            session("01a02d09-4d21-7f0a-9b55-000000000001", "first", "", ""),
            session("01a02d09-4d21-7f0a-9b55-000000000002", "second", "", ""),
        ];
        let mut input = Cursor::new("2\n");
        let mut output = Vec::new();

        let selected =
            choose_from_sessions(&sessions, &mut input, &mut output).expect("choose session");

        assert_eq!(selected.expect("selection").id, sessions[1].id);
        let rendered = String::from_utf8(output).expect("picker output");
        assert!(rendered.contains("saved sessions for this workspace"));
        assert!(rendered.contains("01a02d09-4d"));
        assert!(rendered.contains("choose a number"));
    }

    #[test]
    fn picker_allows_a_new_session_or_rejects_bad_indexes() {
        let sessions = vec![session("aaaa1111-0000-7000-8000-000000000000", "", "", "")];
        let mut output = Vec::new();
        assert!(
            choose_from_sessions(&sessions, &mut Cursor::new("\n"), &mut output)
                .expect("new session choice")
                .is_none()
        );
        let error = choose_from_sessions(&sessions, &mut Cursor::new("0\n"), &mut Vec::new())
            .expect_err("invalid index");
        assert!(error.to_string().contains("between 1 and 1"));
    }

    #[test]
    fn session_description_includes_state_and_title() {
        let mut info = session("aaaa1111-0000-7000-8000-000000000000", "SSE fix", "", "");
        info.locked = true;
        let (label, description) = describe_session(&info, "aaaa1111", false);

        assert_eq!(label, "aaaa1111");
        for expected in ["SSE fix", "4 msg", "open elsewhere"] {
            assert!(description.contains(expected), "{description}");
        }
    }
}
